borgmatic/commands/arguments.py

```diff
@@ -77,13 +95,34 @@ def parse_subparser_arguments(unparsed_arguments, subparsers):
 
     remaining_arguments = list(unparsed_arguments)
 
-    # Now ask each subparser, one by one, to greedily consume arguments.
-    for subparser_name, subparser in subparsers.items():
+    # Now ask each subparser, one by one, to greedily consume arguments, from last to first. This
+    # allows subparsers to consume arguments before their parent subparsers do.
+    remaining_subparser_arguments = []
+
+    # Now ask each subparser, one by one, to greedily consume arguments, from last to first. This
+    # allows subparsers to consume arguments before their parent subparsers do.
+    for subparser_name, subparser in reversed(subparsers.items()):
         if subparser_name not in arguments.keys():
             continue
 
         subparser = subparsers[subparser_name]
-        unused_parsed, remaining_arguments = subparser.parse_known_args(remaining_arguments)
+        unused_parsed, remaining = subparser.parse_known_args(
+            [argument for argument in unparsed_arguments if argument != subparser_name]
+        )
+        remaining_subparser_arguments.append(remaining)
+
+    # Determine the remaining arguments that no subparsers have consumed.
+    if remaining_subparser_arguments:
+        remaining_arguments = [
+            argument
+            for argument in dict.fromkeys(
+                itertools.chain.from_iterable(remaining_subparser_arguments)
+            ).keys()
+            if all(
+                argument in subparser_arguments
+                for subparser_arguments in remaining_subparser_arguments
+            )
+        ]
```

Comment:
IMO this would benefit from some explicit test coverage. I think the easiest way to do that would be to factor it out into a separate function, mock that function within tests of this `parse_subparser_arguments()`, and then independently test the function by passing in various `remaining_subparser_arguments` and asserting the result.

But if you'd rather keep it all inline, you could probably still test the logic by calling `parse_subparser_arguments()` and manipulating a mocked `parse_known_args()` to return the results you want to test.

Let me know though if either approach looks like it will be problematic.

Reply:
Shouldn't this be considered tested, considering failing tests prompted us to rewrite this section ( the extract/list thing)? What should I test here?

Reply:
You're right that it's got _coverage_, meaning the code is exercised by tests, but it's not tested on its own. The code is fairly complex and while it definitely handles the `extract/list` case as you point out, without targeted tests I wouldn't necessarily be convinced it'd handle other cases of finding unique remaining arguments across all subparsers. If this were my feature (and to be clear, it's most definitely not!), I'd probably write a targeted test that, given some contrived `remaining_subparser_arguments`, assert that the code produces the expected `remaining_arguments`. And I might even throw a couple of different variants at it to make sure it handles edge cases like zero overlapping remaining arguments across the subparsers, an empty `remaining_subparser_arguments` element for one subparser, etc etc. Basically whenever I'm scared by complexity I throw test cases at the problem until I'm a little calmer.

Having said all that, this is _your_ feature, so if you're comfortable with the existing coverage, I can be fine with it too.

Reply:
Yeah, I guess it could be tested more.

I tried this - [`6475345` (#71)](https://github.com/borgmatic-collective/borgmatic/pull/71/commits/6475345a8fbaa9b3d9aaeb7444075829c49343ab#diff-d39d71411512f243818b758012a11063b00ebfa284f762c22eca562d30d0bce8R180) to parameterize and separate the unit to test it.

But it's probably wrong since whatever I pass as the array here:

Reply:
Another case that I tried but don't know if this is correct (it passes the test):
```python
        (
            {'extract': flexmock(), 'list': flexmock()},
            ['--progress', '--archive', 'myarchive'],
            {
                'extract': flexmock(parse_known_args=lambda arguments: (flexmock(), ['--archive', 'myarchive'])),
                'list': flexmock(parse_known_args=lambda arguments: (flexmock(), [])),
            },
            [],
        )
```

Reply:
Nevermind, I think I found the code in question. I'll have a look.

Reply:
> From https://github.com/borgmatic-collective/borgmatic/pull/71#issuecomment-1577651275, here's some potential test case data:

Note that my example test case data from above is only for the second half of your factored-out `parse_subparser_arguments()`/`parse_known_subparser_arguments` function. (I hadn't found the actual function before I commented.) So if you wanted a test case for the whole function, it would be like one of your existing examples except `unparsed_arguments` might be `('--latest', 'archive', 'check', 'prune', 'extract', 'list', '--wtf')` with the subparsers set up to mimic the actual `check`, `prune`, `extract`, `list`, etc. But maybe your existing `extract`/`list`-only example would be sufficient.

Reply:
> Another case that I tried but don't know if this is correct (it passes the test):

Okay, finally viewing this test data in light of the actual code. My thoughts:

 * Do you need the action names in the unparsed arguments?
 * Does the `extract` subparser need to consume `--archive` like `list` does?
 * Should the `list` subparser return `--progress` among their remaining arguments? AFAIK it doesn't touch `--progress`.

If you want to "cheat" here, just instrument the actual code with print statements and then run borgmatic with the flags of your test case. Then look at the printed values and steal them for your tests.